File: Data Structures & Algorithms/minimum-recolors-to-get-k-consecutive-black-blocks/submission-5.py
class Solution:
    def minimumRecolors(self, blocks: str, k: int) -> int:
        currentWhites = 0

        # first window of K
        for i in range(k):
            if blocks[i] == "W":
                currentWhites += 1
        
        minOps = currentWhites

        for i in range(k, len(blocks)):
            if blocks[i]== "W":
                currentWhites += 1

            if blocks[i - k] == "W":
                currentWhites -= 1

            minOps = min(minOps, currentWhites)

        return minOps

# Time Complexity = O(N)
# Space Complexity 0(1)
        

File: Data Structures & Algorithms/contains-duplicate-ii/submission-5.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {boolean}
     */
    containsNearbyDuplicate(nums, k) {

        const map = new Map();

        for(let i=0; i<nums.length; i++){
            let num = nums[i];

            if(map.has(num)){
                let dup = Math.abs(map.get(num) - i)
                if(dup <= k){
                    return true
                }
                map.set(num, i)
            } else {
                map.set(num, i)
            }
        }
        
        return false
    }
}

//Time Complexity: O(N)
//Space Complexity: O(N)
//This solution does not make use of sliding window. Check solution 2 for sliding window.
